libcxx/utils/generate_feature_test_macro_components.py

```diff
@@ -1857,6 +1858,215 @@ def produce_docs():
+                "number": "P0220R1",
+
+                # optional
+                # Contains the title of the paper that is part of the FTM
+                # version.
+                "title": "Adopt Library Fundamentals V1 TS Components for C++17"
+
+                # required
+                # The implementation status of the paper.
+                "implemented": true
+              }
+            ]
+          }
+        },
+
+        # required
+        # A sorted list of headers that should provide the FTM. The header
+        # <version> is automatically added to this list. This list could be
+        # empty. For example, __cpp_lib_modules is only present in version.
+        # Requiring the field makes it easier to detect accidental omission.
+        "headers": [
+          "any"
+        ],
+
+        # optional, required when libcxx_guard is present
+        # This field is used only to generate the unit tests for the
+        # feature-test macros. It can't depend on macros defined in <__config>
+        # because the `test/std/` parts of the test suite are intended to be
+        # portable to any C++ standard library implementation, not just libc++.
+        # It may depend on
+        # * macros defined by the compiler itself, or
+        # * macros generated by CMake.
+        # In some cases we add also depend on macros defined in
+        # <__availability>.
+        "test_suite_guard": "!defined(_LIBCPP_VERSION) || _LIBCPP_AVAILABILITY_HAS_PMR"
+
+        # optional, required when test_suite_guard is present
+        # This field is used only to guard the feature-test macro in
+        # <version>. It may be the same as `test_suite_guard`, or it may
+        # depend on macros defined in <__config>.
+        "libcxx_guard": "_LIBCPP_AVAILABILITY_HAS_PMR"
+      },
+    ]
+    """
+
+    # The JSON data structor.
+    __data = None
+
+    # These values are used internally multiple times. They are lazily loaded
+    # and cached. Values that are expected to be used once are not cached.
+    __std_dialects = None
+    __std_dialect_versions = None
+    __dialect_versions = None
+
+    def __init__(self, filename):
+        """Initializes the class with the JSON data in the file 'filename'."""
+        self.__data = json.load(open(filename))
+
+    def get_std_dialects(self):
+        """Returns the C++ dialects avaiable.
+
+        The available dialects are based on the 'c++xy' keys found the 'values'
+        entries in '__data'. So when WG21 starts to feature-test macros for a
+        future C++ Standard this dialect will automatically be available.
+
+        The return value is a sorted list with the C++ dialects used. Since FTM
+        were added in C++14 the list will not contain C++98 or C++11.
+        """
+        if not self.__std_dialects:
+            self.__std_dialects = get_std_dialects(self.__data)
+
+        return self.__std_dialects
+
+    def get_std_dialect_versions(self):
```

Comment:
We could instead provide the following interface:

```python
def get_standard_ftms(self, dialect: str) -> List[Tuple[str, int]]:
  # returns the list of ('__cpp_lib_foo', 2017xxx) FTMs in the Standard

def get_implemented_ftms(self, dialect: str) -> List[Tuple[str, int]]:
  # returns the list of ('__cpp_lib_foo', 2017xxx) FTMs implemented by libc++
```

Reply:
I renamed the functions, including the removal of `get_` since they became properties. We don't use types in this file so I did not add them. I wonder how useful they are since AFAIK they are never validated in LLVM.

Reply:
^ As discussed, I think these annotations make it much easier to understand APIs so we should move towards using them.